Chapter 6 (#6)
Andarist/total-typescript-book: book-content/chapters/06-objects.md

```typescript
const wishYouWereHereSales: AlbumSales = {
  title: "Wish You Were Here",
  artist: "Pink Floyd",
  releaseYear: 1975
  unitsSold: 13000000,
  revenue: 65000000,
};
```

If the contract of the `AlbumSales` type isn't fulfilled when creating a new object, TypeScript will raise an error.

It's also possible to intersect more than two types:

```typescript
type AlbumSales = Album & SalesData & { genre: string };
```

This is a useful method for creating new types from existing ones.

#### Intersection Types With Primitives

It's worth noting that intersection types can also be used with primitives, like `string` and `number` - though it often produces odd results.

For instance, let's try intersecting `string` and `number`:

```typescript
type StringAndNumber = string & number;
```

What type do you think `StringAndNumber` is? It's actually `never`. This is because `string` and `number` have innate properties that can't be combined together.

This also happens when you intersect two object types with an incompatible property:

```ts twoslash
type User1 = {
  age: number;
};

type User2 = {
  age: string;
};

type User = User1 & User2;
//   ^?
```

> In this case, the `age` property resolves to `never` because it's impossible for a single property to be both a `number` and a `string`.

Author: but it's only the `age` property that resolves to `never` - the type itself, its potential other properties, etc wont be never-ified. That makes the previous definition that you coined above inaccurate:
> What type do you think `StringAndNumber` is? It's actually `never`. This is because `string` and `number` have innate properties that can't be combined together.

In that case it's not about mismatched properties but about completely disjoint domains (or something ;p). Like, there is maybe some overlap between both of those that could be the output of such operation. For example, both `number` and `string` have `{ toString(): string }`. For practical purposes, it's rather useless to create an object with only compatible methods though.

  artist: "Grateful Dead",
  releaseYear: 1991,
  concertVenue: "Great American Music Hall",
  concertDate: new Date("1975-08-13"),
};
```

Just as adding additional `&` operators add to an intersection, it's also possible for an interface to extend multiple other interfaces by separating them with commas:

```typescript
interface BoxSet extends StudioAlbum, LiveAlbum {
  numberOfDiscs: number;
}
```

### Intersections vs `interface extends`

We've now covered two separate TypeScript syntaxes for extending object types: `&` and `interface extends`. So, which is better?

You should choose `interface extends` for two reasons.

#### Better Errors When Merging Incompatible Types

We saw earlier that when you intersect two object types with an incompatible property, TypeScript will resolve the property to `never`:

```typescript
type User1 = {
  age: number;
};

type User2 = {
  age: string;
};

type User = User1 & User2;
```

When using `interface extends`, TypeScript will raise an error when you try to extend an interface with an incompatible property:

```ts twoslash
// @errors: 2430
interface User1 {
  age: number;
}

interface User extends User1 {
  age: string;
}
```

> This is very different because it actually sources an error. With intersections, TypeScript will only raise an error when you try to access the `age` property, not when you define it.

Author: > With intersections, TypeScript will only raise an error when you try to access the `age` property, not when you define it.

It won't even raise an error - it will just resolve it to `never` and since `never` is assignable to everything... this might *silently* creep issues into ur code.

So, `interface extends` is better for catching errors when building out your types.

#### Better TypeScript Performance

When you're working in TypeScript, the performance of your types should be at the back of your mind. In large projects, how you define your types can have a big impact on how fast your IDE feels, and how long it takes for `tsc` to check your code.

> `interface extends` is much better for TypeScript performance than intersections. With intersections, the intersection is recomputed every time it's used. This can be slow, especially when you're working with complex types.

Author: Intersections are heavily cached too. Some resolutions sometimes (maybe? I'm not sure) can be recomputed but in many cases, they won't.

Interfaces are faster. TypeScript can cache the resulting type of an interface based on its name. So if you use `interface extends`, TypeScript only has to compute the type once, and then it can reused it every time you use the interface.

> #### Conclusion

Author: a third reason could be that the IDE preserves the interfaces' names more often than not whereas with type aliases u can often see the content (the intersection) and that might lead to gnarly error messages etc

### Combining Known and Dynamic Keys

In many cases there will be a base set of keys we know we want to include, but we also want to allow for additional keys to be added dynamically.

For example, say we are working with a base set of awards we know were nominations, but we don't know what other awards are in play. We can use the `Record` type to define a base set of awards and then use an intersection to extend it with an index signature for additional awards:

```typescript
type BaseAwards = "Grammy" | "MercuryPrize" | "Billboard";

type ExtendedAlbumAwards = Record<BaseAwards, boolean> & {
  [award: string]: boolean;
};

const extendedNominations: ExtendedAlbumAwards = {
  Grammy: true,
  MercuryPrize: false,
  Billboard: true, // Additional awards can be dynamically added
  "American Music Awards": true,
};
```

This technique would also work when using an interface and the `extends` keyword:

```typescript
interface BaseAwards {
  Grammy: boolean;
  MercuryPrize: boolean;
  Billboard: boolean;
}

interface ExtendedAlbumAwards extends BaseAwards {
  [award: string]: boolean;
}
```

> This version is preferable because, in general, `interface extends` is preferable to intersections.

Author: it's even more important with index signatures because to make sound types all of ur values have to be compatible with applicable index signatures. Intersections won't warn you about this at all.

People often ask to be able to break this rule but that just creates problems: [TS playground](https://www.typescriptlang.org/play/?#code/CYUwxgNghgTiAEYD2A7AzgF3kgRgKwC54BveAbQGsjMYBLFAcwF0ickkIQoV4BfeAGQl4FFEgDuKIigCuAWxwgYfANwAoNcnRY4aGRAwBGeAF5s+AHSiJKdQHo78J-AB6Afg1bMIsZOoY6RlN4ACJrSRD1Lx0QPQMAJmDcPEpfFCZ7R2d3NSA)

// inside lib.es5.d.ts
declare type PropertyKey = string | number | symbol;
```

Because `PropertyKey` works with all possible keys, it's great for working with dynamic keys where you aren't sure what the type of the key will be.

For example, when using an index signature you could set the key type to `PropertyKey` in order to allow for any valid key type:

```typescript
type Album = {
  [key: PropertyKey]: string;
};
```

### `object`

Similar to `string`, `number`, and `boolean`, `object` is a global type in TypeScript.

It represents more types than you might expect. Instead of representing only objects like `{}` or `new Object()`, `object` represents any non-primitive type. This includes arrays, functions, and objects.

So a function like this:

```typescript
function acceptAllNonPrimitives(obj: object) {}
```

Would accept any non-primitive value:

```typescript
acceptAllNonPrimitives({});
acceptAllNonPrimitives([]);
acceptAllNonPrimitives(() => {});
```

But error on primitives:

```ts twoslash
// @errors: 2345
function acceptAllNonPrimitives(obj: object) {}

// ---cut---
acceptAllNonPrimitives(1);
acceptAllNonPrimitives("hello");
acceptAllNonPrimitives(true);
```

> This means that the `object` type is rarely useful by itself. Using `Record` is usually a better choice. For instance, if you want to accept any object type, you can use `Record<string, unknown>`.

Author: This is a fallacy - to the best of my knowledge there is no way clear way to create a type that would represent "a plain object". `Record<string, unknown>` has its gotchas too and it just doesn't deliver that: [TS playground](https://www.typescriptlang.org/play/?#code/JYOwLgpgTgZghgYwgAgGIHt3IN4ChnIBGcUAXMgM5hSgDmA3LgL64LohXIybkZYC8OfERLkARGIA0zRrhgBXEAjDB2yREgAOYAPKEAVhGUUAFOgPkASkfRQAJgB4qNELUnJFAaxDoA7iAA+AEocFlwNCG09Q2MTbnQgxiA)

Author: it only kinda works with objects defined `type` because those have "inferrable index signatures" - which is far from obvious

};

// ---cut---
const obj = {
  1: "bar",
};
```

Because an object can also have a symbol as a key, there is also a test for that case. It currently has type errors for `fooSymbol` and `barSymbol` when calling `hasKey`:

```ts twoslash
// @lib: dom,es2023,dom.iterable
// @errors: 2345
const hasKey = (obj: object, key: string) => {
  return obj.hasOwnProperty(key);
};

// ---cut---
const fooSymbol = Symbol("foo");
const barSymbol = Symbol("bar");

const obj = {
  [fooSymbol]: "bar",
};
```

Your task is to update the `hasKey` function so that all of these tests pass. Try to be as concise as possible!

<Exercise title="Exercise 4: Dynamic Key Support" filePath="/src/020-objects/086-property-key-type.problem.ts"></Exercise>

#### Solution 1: Use an Index Signature for Dynamic Keys

Here are the three solutions:

You can use an inline index signature:

```typescript
const scores: {
  [key: string]: number;
} = {};
```

Or an interface:

```typescript
interface Scores {
  [key: string]: number;
}
```

Or a type:

```typescript
type Scores = {
  [key: string]: number;
};
```

Or finally, a record:

```typescript
const scores: Record<string, number> = {};
```

> #### Solution 2: Default Properties with Dynamic Keys

Author: calling them default feels weird to me - i'd use known/concrete

```typescript
interface RequiredScores {
  math: number;
  english: number;
  science: number;
}

interface Scores extends RequiredScores {
  [key: string]: number;
}
```

These two are functionally equivalent, except for the fact that you get access to the `RequiredScores` interface if you need to use that seprately.

#### Solution 3: Restricting Object Keys

##### A Failed First Attempt at Using Record

We know that the values of the `Configurations` object will be `apiBaseUrl`, which is a string, and `timeout`, which is a number.

It may be tempting to use a Record to set the key as a string and the value an object with the properties `apiBaseUrl` and `timeout`:

```typescript
type Configurations = Record<
  string,
  {
    apiBaseUrl: string;
    timeout: number;
  }
>;
```

However, having the key as `string` still allows for the `notAllowed` key to be added to the object. We need to make the keys dependent on the `Environment` type.

##### The Correct Approach

Instead, we can specify the `key` as `Environment` inside the Record:

```typescript
type Configurations = Record<
  Environment,
  {
    apiBaseUrl: string;
    timeout: number;
  }
>;
```

> Now TypeScript will throw an error when the object includes a key that doesn't exist in `Environment`, like `notAllowed`.

Author: This is not the only place when a language like this is used but I wonder if it wouldn't be better to adjust all of them to use "report an error" over "throw an error".

### `Partial`

The Partial utility type lets you create a new object type from an existing one, except all of its properties are optional.

Consider an Album interface that contains detailed information about an album:

```typescript
interface Album {
  id: number;
  title: string;
  artist: string;
  releaseYear: number;
  genre: string;
}
```

When we want to update an album's information, we might not have all the information at once. For example, it can be difficult to decide what genre to assign to an album before it's released.

Using the `Partial` utility type and passing in `Album`, we can create a type that allows us to update any subset of an album's properties:

```typescript
type PartialAlbum = Partial<Album>;
```

Now we have a `PartialAlbum` type where `id`, `title`, `artist`, `releaseYear`, and `genre` are all optional.

This means we can create a function which only receives a subset of the album's properties:

```typescript
const updateAlbum = (album: PartialAlbum) => {
  // ...
};

updateAlbum({ title: "Geogaddi", artist: "Boards of Canada" });
```

> ### `Required`

Author: it might be neat to mention somewhere that Partial/Required are not limited to objects - they work fine with tuples as well :)

```typescript
// Expected:
type MusicProductWithoutId =
  | Omit<Album, "id">
  | Omit<CollectorEdition, "id">
  | Omit<DigitalRelease, "id">;

// Actual:
type MusicProductWithoutId = {
  title: string;
};
```

This is particularly annoying given that `Partial` and `Required` work as expected with union types:

```typescript
type PartialMusicProduct = Partial<MusicProduct>;

// Hovering over PartialMusicProduct shows:
type PartialMusicProduct =
  | Partial<Album>
  | Partial<CollectorEdition>
  | Partial<DigitalRelease>;
```

This stems from how `Omit` processes union types. Rather than iterating over each union member, it amalgamates them into a single structure it can understand.

The technical reason for this is that `Omit` and `Pick` are not distributive. This means that when you use them with a union type, they don't operate individually on each union member.

#### The `DistributiveOmit` and `DistributivePick` Types

In order to address this, we can create a `DistributiveOmit` type. It's defined similarly to Omit but operates individually on each union member. Note the inclusion of `PropertyKey` in the type definition to allow for any valid key type:

```typescript

> type DistributiveOmit<T, K extends PropertyKey> = T extends any
>   ? Omit<T, K>
>   : never;

Author: or :p
```suggestion
type DistributiveOmit<T, K extends PropertyKey> = { [P in keyof T as Exclude<P, K>]: T[P]; }
```

```typescript
const fetchUser = async (): Promise<PickedUser> => {
  ...
```

##### Using Omit

The Omit utility type can also be used to create a new type that excludes the `id` and `role` properties from the `User` interface:

```typescript
type OmittedUser = Omit<User, "id" | "role">;
```

Then the `fetchUser` function can be updated to return a `Promise` of `OmittedUser`:

```typescript
const fetchUser = async (): Promise<OmittedUser> => {
  ...
```

##### Extending an Interface

We could create an interface `NameAndEmail` that contains a `name` and `email` property, along with updating the `User` interface to remove those properties in favor of extending them:

```typescript
interface NameAndEmail {
  name: string;
  email: string;
}

interface User extends NameAndEmail {
  id: string;
  role: string;
}
```

Then the `fetchUser` function could return a `Promise` of `NameAndEmail`:

```typescript
const fetchUser = async (): Promise<NameAndEmail> => {
  // ...
};
```

`Omit` will mean that the object grows as the source object grows. `Pick` and `interface extends` will mean that the object will stay the same size. So depending on requirements, you can choose the best approach.

> #### Solution 2: Updating a Product

Author: Usually, your Solution X were interchangeable but, in here, this one feels like an extension of the previous one (like an extra step). The previous one doesn't mention `Partial` at all (which is a requirement here).

Author: hm, maybe this just isn't true? I see now that Solution X is meant to correspond to Excercise X. I was under impression that before you were using multiple Solutions to a single Excercise at times but maybe those were always presented under subheaders or smth
